### PR TITLE
[air/output] Use Tune reporter if Tuner/tune.run was used

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Currently we use the number of samples to detect if we should use the Train output reporter or the tune output reporter.

However, this is confusing when developing on a small scale, e.g. a tuning run with only 1 trial, before going to multiple trials. Instead, we should use the entrypoint to decide which reporter we use. The `Trainer` will use the Train reporter, and Tuner, tune.run, and tune.run_experiments will use the tune reporter.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by scottsun94** [comment]
This change makes sense!

One question:
how about Rllib? As long as it doesn't use `tuner`, `tune.run`, and `tune.run_experiments`, it should just use `trainer`?

**Comment by justinvyu** [comment]
> This change makes sense!
> 
> One question: how about Rllib? As long as it doesn't use `tuner`, `tune.run`, and `tune.run_experiments`, it should just use `trainer`?

@scottsun94 RLlib can only run through the Tune entrypoints. Trainer is from Ray Train.
Training an RLlib `Algorithm` outside of Tune doesn't come with reporting capabilities.

**Review by justinvyu — COMMENTED** [review]
Makes sense! Are the failing tests unrelated?

**Comment by krfricke** [comment]
> Makes sense! Are the failing tests unrelated?

Yes, they're due to a breakage that was fixed in #36157. I'll merge master to reflect this in the CI.

File: python/ray/tune/experimental/output.py [hunk 1]
```diff
@@ -601,13 +602,18 @@ def _print_heartbeat(self, trials, *args, force: bool = False):
...
     config: Optional[Dict] = None,
 ):
     # TODO: Add JupyterNotebook and Ray Client case later.
     rich_enabled = bool(int(os.environ.get("RAY_AIR_RICH_LAYOUT", "0")))
-    if num_samples and num_samples > 1:
+    if entrypoint in {
+        AirEntrypoint.TUNE_RUN,
+        AirEntrypoint.TUNE_RUN_EXPERIMENTS,
+        AirEntrypoint.TUNER,
+    } or (num_samples and num_samples > 1):
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/experimental/output.py:616]: SG, will update the PR

**Comment by scottsun94** [comment]
> > This change makes sense!
> 
> > 
> 
> > One question: how about Rllib? As long as it doesn't use `tuner`, `tune.run`, and `tune.run_experiments`, it should just use `trainer`?
> 
> 
> 
> @scottsun94 RLlib can only run through the Tune entrypoints. Trainer is from Ray Train.
> 
> Training an RLlib `Algorithm` outside of Tune doesn't come with reporting capabilities.

This makes sense. But what's the story of output for rllib? They need to build their owner output  reporter?

**Comment by krfricke** [comment]
> > > This change makes sense!
> > 
> > 
> > > 
> > 
> > 
> > > One question: how about Rllib? As long as it doesn't use `tuner`, `tune.run`, and `tune.run_experiments`, it should just use `trainer`?
> > 
> > 
> > @scottsun94 RLlib can only run through the Tune entrypoints. Trainer is from Ray Train.
> > Training an RLlib `Algorithm` outside of Tune doesn't come with reporting capabilities.
> 
> This makes sense. But what's the story of output for rllib (training an RLlib Algorithm outside of Tune)? Rllib needs to build its own output reporter?

Yes, if rllib is not using tune/air for execution, they have to implement their own reporting logic. They can potentially still use the same output reporter, but they have to call the hooks themselves.

The logic from this PR would then still use the train (single run) reporter, so it should be unaffected by the changes in this PR.

**Review by justinvyu — APPROVED** [review]

**Comment by scottsun94** [comment]
> > > > This change makes sense!
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > One question: how about Rllib? As long as it doesn't use `tuner`, `tune.run`, and `tune.run_experiments`, it should just use `trainer`?
> > > 
> > > 
> > > @scottsun94 RLlib can only run through the Tune entrypoints. Trainer is from Ray Train.
> > > Training an RLlib `Algorithm` outside of Tune doesn't come with reporting capabilities.
> > 
> > 
> > This makes sense. But what's the story of output for rllib (training an RLlib Algorithm outside of Tune)? Rllib needs to build its own output reporter?
> 
> Yes, if rllib is not using tune/air for execution, they have to implement their own reporting logic. They can potentially still use the same output reporter, but they have to call the hooks themselves.
> 
> The logic from this PR would then still use the train (single run) reporter, so it should be unaffected by the changes in this PR.

It makes sense. Thanks!

**Pull request merged**